File: Exercises/Exercise04/java_dgemm/src/benchmark/DoubleDgemm.java
package benchmark;

public class DoubleDgemm implements IDgemm{
    private double alpha;
    private double beta;
    private double[][] matrixA;
    private double[][] matrixB;
    private double[][] matrixC;
    Reader reader;

    public DoubleDgemm(Reader reader){
        this.reader = reader;
        convertValues();
    }
    @Override
    public void calculateDgemm(){
        matrixC = addMatrices(multiplyMatrixByMatrix(multiplyScalarByMatrix(alpha, matrixA), matrixB),
                multiplyScalarByMatrix(beta, matrixC));
    }
    @Override
    public String[] getAnswer(){
        return convertMatrixToString(matrixC);
    }

    public double[][] multiplyMatrixByMatrix(double[][] firstMat, double[][] secondMat){
        double[][] resultMatrix = new double[firstMat.length][secondMat[0].length];

        for (int m = 0; m < firstMat.length; m++){
            for (int n = 0; n < secondMat[0].length; n++){
                for (int k = 0; k < secondMat.length; k++){
                    resultMatrix[m][n] += firstMat[m][k] * secondMat[k][n];
                }
            }
        }
        return resultMatrix;
    }

    public double[][] multiplyScalarByMatrix(double scal, double[][] mat){
        for (int m = 0; m < mat.length; m++){
            for (int n = 0; n < mat[0].length; n++){
                mat[m][n] = mat[m][n] * scal;
            }
        }
        return mat;
    }

    public double[][] addMatrices(double[][] firstMat, double[][] secondMat){
        for (int m = 0; m < firstMat.length; m++){
            for (int n = 0; n < secondMat[0].length; n++){
                firstMat[m][n] += secondMat[m][n];
            }
        }
        return firstMat;
    }

    public void convertValues(){
        try {
            int size = reader.getMatrixSize();
            matrixA = new double[size][size];
            matrixB = new double[size][size];
            matrixC = new double[size][size];

            alpha = Double.parseDouble(reader.getAlpha());
            beta = Double.parseDouble(reader.getBeta());
            matrixA = convertMatrixToDouble(reader.getMatrixA());
            matrixB = convertMatrixToDouble(reader.getMatrixB());
            matrixC = convertMatrixToDouble(reader.getMatrixC());
        } catch (NumberFormatException e) {
            System.err.println("Error: " + e.getMessage());
        }

    }

    public double[][] convertMatrixToDouble(String[][] strMatrix){
        double[][] matrix = new double[strMatrix.length][strMatrix.length];
        try {
            for (int i = 0; i < strMatrix.length; i++){
                for (int j = 0; j < strMatrix.length; j++){
                    matrix[i][j] = Double.parseDouble(strMatrix[i][j]);
                }
            }
        } catch (NumberFormatException e) {
            System.err.println("Error: " + e.getMessage());
        }
        return matrix;
    }

    public String[] convertMatrixToString(double[][] matrix){
        String[] result = new String[matrix.length];
        for (int i = 0; i < matrix.length; i++){
            StringBuilder string = new StringBuilder();
            for (int j = 0; j < matrix.length; j++){
                string.append(matrix[i][j] + " ");
            }
            result[i] = string.toString().trim();
        }
        return result;
    }

}

File: Exercises/Exercise04/Csharp.Dgemm/Csharp.Dgemm/DoubleDgemm.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Csharp.Dgemm
{
    public class DoubleDgemm : IDgemm
    {
        private double alpha;
        private double beta;
        private double[,] matrixA;
        private double[,] matrixB;
        private double[,] matrixC;
        Reader reader;

        public DoubleDgemm(Reader reader)
        {
            this.reader = reader;
            ConvertValues();          
        }

        public void ConvertValues()
        {
            System.Threading.Thread.CurrentThread.CurrentCulture = new System.Globalization.CultureInfo("en-US");
            try
            {
                matrixA = new double[reader.MatrixSize, reader.MatrixSize];
                matrixB = new double[reader.MatrixSize, reader.MatrixSize];
                matrixC = new double[reader.MatrixSize, reader.MatrixSize];

                alpha = Double.Parse(reader.Alpha);
                beta = Double.Parse(reader.Beta);
                matrixA = ConvertMatrixToDouble(reader.MatrixA);
                matrixB = ConvertMatrixToDouble(reader.MatrixB);
                matrixC = ConvertMatrixToDouble(reader.MatrixC);
            }
            catch (InvalidCastException e)
            {
                Console.WriteLine(e.Message);
            }
        }

        public double[,] ConvertMatrixToDouble(string[,] strMatrix)
        {
            double[,] matrix = new double[reader.MatrixSize, reader.MatrixSize];
            try
            {
                for (int i = 0; i < reader.MatrixSize; i++)
                {
                    for (int j = 0; j < reader.MatrixSize; j++)
                    {
                        matrix[i, j] = Double.Parse(strMatrix[i, j]);
                    }
                }
            }
            catch (InvalidCastException e)
            {
                Console.WriteLine(e.Message);
            }
            return matrix;
        }

        public string[] GetAnswer()
        {
            return ConvertMatrixToString(matrixC);
        }

        public string[] ConvertMatrixToString(double[,] matrix)
        {
            string[] result = new string[reader.MatrixSize];
            for (int i = 0; i < reader.MatrixSize; i++)
            {
                StringBuilder sr = new StringBuilder();
                for (int j = 0; j < reader.MatrixSize; j++)
                {
                    sr.Append(matrix[i, j] + " ");
                }
                result[i] = sr.ToString().Trim();
            }
            return result;
        }

        public void CalculateDgemm()
        {
            matrixC = AddMatrices(MultiplyMatrixByMatrix(MultiplyScalarByMatrix(alpha, matrixA), matrixB), MultiplyScalarByMatrix(beta, matrixC));           
        }


        public double[,] MultiplyMatrixByMatrix(double[,] firstMat, double[,] secondMat)
        {
            double[,] resultMatrix = new double[reader.MatrixSize, reader.MatrixSize];

            for (int m = 0; m < reader.MatrixSize; m++)
            {
                for (int n = 0; n < reader.MatrixSize; n++)
                {
                    for (int k = 0; k < reader.MatrixSize; k++)
                    {
                        resultMatrix[m, n] += firstMat[m, k] * secondMat[k, n];
                    }
                }
            }

            return resultMatrix;
        }

        public double[,] MultiplyScalarByMatrix(double scal, double[,] mat)
        {
            for (int m = 0; m < reader.MatrixSize; m++)
            {
                for (int n = 0; n < reader.MatrixSize; n++)
                {
                    mat[m, n] = mat[m, n] * scal;
                }
            }
            return mat;
        }

        public double[,] AddMatrices(double[,] firstMat, double[,] secondMat)
        {
            for (int m = 0; m < reader.MatrixSize; m++)
            {
                for (int n = 0; n < reader.MatrixSize; n++)
                {
                    firstMat[m, n] += secondMat[m, n];
                }
            }
            return firstMat;
        }
    }
}

